File: examples3d/keva3.rs
use rapier_testbed3d::Testbed;
use rapier3d::prelude::*;

pub fn build_block(
    testbed: &mut Testbed,
    bodies: &mut RigidBodySet,
    colliders: &mut ColliderSet,
    half_extents: Vector<f32>,
    shift: Vector<f32>,
    (mut numx, numy, mut numz): (usize, usize, usize),
) {
    let dimensions = [half_extents.xyz(), half_extents.zyx()];
    let block_width = 2.0 * half_extents.z * numx as f32;
    let block_height = 2.0 * half_extents.y * numy as f32;
    let spacing = (half_extents.z * numx as f32 - half_extents.x) / (numz as f32 - 1.0);
    let mut color0 = [0.7, 0.5, 0.9];
    let mut color1 = [0.6, 1.0, 0.6];

    for i in 0..numy {
        std::mem::swap(&mut numx, &mut numz);
        let dim = dimensions[i % 2];
        let y = dim.y * i as f32 * 2.0;

        for j in 0..numx {
            let x = if i % 2 == 0 {
                spacing * j as f32 * 2.0
            } else {
                dim.x * j as f32 * 2.0
            };

            for k in 0..numz {
                let z = if i % 2 == 0 {
                    dim.z * k as f32 * 2.0
                } else {
                    spacing * k as f32 * 2.0
                };

                // Build the rigid body.
                let rigid_body = RigidBodyBuilder::dynamic().translation(vector![
                    x + dim.x + shift.x,
                    y + dim.y + shift.y,
                    z + dim.z + shift.z
                ]);
                let handle = bodies.insert(rigid_body);
                let collider = ColliderBuilder::cuboid(dim.x, dim.y, dim.z);
                colliders.insert_with_parent(collider, handle, bodies);

                testbed.set_initial_body_color(handle, color0);
                std::mem::swap(&mut color0, &mut color1);
            }
        }
    }

    // Close the top.
    let dim = half_extents.zxy();

    for i in 0..(block_width / (dim.x * 2.0)) as usize {
        for j in 0..(block_width / (dim.z * 2.0)) as usize {
            // Build the rigid body.
            let rigid_body = RigidBodyBuilder::dynamic().translation(vector![
                i as f32 * dim.x * 2.0 + dim.x + shift.x,
                dim.y + shift.y + block_height,
                j as f32 * dim.z * 2.0 + dim.z + shift.z
            ]);
            let handle = bodies.insert(rigid_body);
            let collider = ColliderBuilder::cuboid(dim.x, dim.y, dim.z);
            colliders.insert_with_parent(collider, handle, bodies);
            testbed.set_initial_body_color(handle, color0);
            std::mem::swap(&mut color0, &mut color1);
        }
    }
}

pub fn init_world(testbed: &mut Testbed) {
    /*
     * World
     */
    let mut bodies = RigidBodySet::new();
    let mut colliders = ColliderSet::new();
    let impulse_joints = ImpulseJointSet::new();
    let multibody_joints = MultibodyJointSet::new();

    /*
     * Ground
     */
    let ground_size = 50.0;
    let ground_height = 0.1;

    let rigid_body = RigidBodyBuilder::fixed().translation(vector![0.0, -ground_height, 0.0]);
    let handle = bodies.insert(rigid_body);
    let collider = ColliderBuilder::cuboid(ground_size, ground_height, ground_size);
    colliders.insert_with_parent(collider, handle, &mut bodies);

    /*
     * Create the cubes
     */
    let half_extents = vector![0.02, 0.1, 0.4] / 2.0 * 10.0;
    let mut block_height = 0.0;
    // These should only be set to odd values otherwise
    // the blocks won't align in the nicest way.
    let numy = [0, 9, 13, 17, 21, 41];
    let mut num_blocks_built = 0;

    for i in (1..=5).rev() {
        let numx = i;
        let numy = numy[i];
        let numz = numx * 3 + 1;
        let block_width = numx as f32 * half_extents.z * 2.0;
        build_block(
            testbed,
            &mut bodies,
            &mut colliders,
            half_extents,
            vector![-block_width / 2.0, block_height, -block_width / 2.0],
            (numx, numy, numz),
        );
        block_height += numy as f32 * half_extents.y * 2.0 + half_extents.x * 2.0;
        num_blocks_built += numx * numy * numz;
    }

    println!("Num keva blocks: {num_blocks_built}");

    /*
     * Set up the testbed.
     */
    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints);
    testbed.look_at(point![100.0, 100.0, 100.0], Point::origin());
}

File: python/rapier-testbed/rapier_testbed/examples3/keva3.py
"""Port of examples3d/keva3.rs (Keva-style tower)."""
from __future__ import annotations

import rapier3d as rp
from .._registry import register

CATEGORY = "Collisions"
NAME = "Keva tower"


def _build_block(bodies, colliders, half_extents, shift, dims):
    """Build a single block of the Keva tower.

    ``half_extents`` is a 3-tuple; ``dims`` is ``(numx, numy, numz)``.
    """
    numx, numy, numz = dims
    dimensions = [
        half_extents,
        (half_extents[2], half_extents[1], half_extents[0]),
    ]
    block_width = 2.0 * half_extents[2] * numx
    block_height = 2.0 * half_extents[1] * numy
    spacing = (half_extents[2] * numx - half_extents[0]) / (numz - 1.0)

    for i in range(numy):
        # mem::swap(&mut numx, &mut numz)
        numx, numz = numz, numx
        dim = dimensions[i % 2]
        y = dim[1] * i * 2.0
        for j in range(numx):
            if i % 2 == 0:
                x = spacing * j * 2.0
            else:
                x = dim[0] * j * 2.0
            for k in range(numz):
                if i % 2 == 0:
                    z = dim[2] * k * 2.0
                else:
                    z = spacing * k * 2.0
                body = rp.RigidBody.dynamic().translation(
                    (x + dim[0] + shift[0], y + dim[1] + shift[1], z + dim[2] + shift[2])
                )
                h = bodies.insert(body)
                colliders.insert_with_parent(
                    rp.Collider.cuboid(dim[0], dim[1], dim[2]), h, bodies
                )

    # Top cap.
    dim = (half_extents[2], half_extents[0], half_extents[1])
    for i in range(int(block_width / (dim[0] * 2.0))):
        for j in range(int(block_width / (dim[2] * 2.0))):
            body = rp.RigidBody.dynamic().translation(
                (
                    i * dim[0] * 2.0 + dim[0] + shift[0],
                    dim[1] + shift[1] + block_height,
                    j * dim[2] * 2.0 + dim[2] + shift[2],
                )
            )
            h = bodies.insert(body)
            colliders.insert_with_parent(
                rp.Collider.cuboid(dim[0], dim[1], dim[2]), h, bodies
            )


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    ground_size = 50.0
    ground_height = 0.1
    ground = rp.RigidBody.fixed().translation((0.0, -ground_height, 0.0))
    ground_h = bodies.insert(ground)
    colliders.insert_with_parent(
        rp.Collider.cuboid(ground_size, ground_height, ground_size), ground_h, bodies
    )

    # `Vector::new(0.02, 0.1, 0.4) / 2.0 * 10.0` → (0.1, 0.5, 2.0).
    half_extents = (0.02 / 2.0 * 10.0, 0.1 / 2.0 * 10.0, 0.4 / 2.0 * 10.0)
    block_height = 0.0
    numy = [0, 9, 13, 17, 21, 41]
    for i in range(5, 0, -1):
        numx = i
        ny = numy[i]
        numz = numx * 3 + 1
        block_width = numx * half_extents[2] * 2.0
        _build_block(
            bodies,
            colliders,
            half_extents,
            (-block_width / 2.0, block_height, -block_width / 2.0),
            (numx, ny, numz),
        )
        block_height += ny * half_extents[1] * 2.0 + half_extents[0] * 2.0

    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    testbed.look_at((100.0, 100.0, 100.0), (0.0, 0.0, 0.0))


register(CATEGORY, NAME, init_world, dim=3)


if __name__ == "__main__":
    from .._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
